src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java

```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
             default: throw new IllegalArgumentException("Unknown datatype '" + datatype + "'");
         }
     }
+
+    private String externalTableName(String tableName, String suffix) {
+        return String.format("%s%s_%s", PDAO_EXTERNAL_TABLE_PREFIX, tableName, suffix);
+    }
+
+    public String createSoftDeleteExternalTable(Dataset dataset, String path, String tableName, String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+        String extTableName = externalTableName(tableName, suffix);
+        TableId tableId = TableId.of(prefixName(dataset.getName()), extTableName);
+        Schema schema = Schema.of(Field.of(PDAO_ROW_ID_COLUMN, LegacySQLTypeName.STRING));
+        ExternalTableDefinition tableDef = ExternalTableDefinition.of(path, schema, FormatOptions.csv());
+        TableInfo tableInfo = TableInfo.of(tableId, tableDef);
+        bigQueryProject.getBigQuery().create(tableInfo);
+        return extTableName;
+    }
+
+    public boolean deleteSoftDeleteExternalTable(Dataset dataset, String tableName, String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+        String extTableName = externalTableName(tableName, suffix);
+        return bigQueryProject.deleteTable(prefixName(dataset.getName()), extTableName);
+    }
+
+    private static final String insertSoftDeleteTemplate =
+        "INSERT INTO `<project>.<dataset>.<softDeleteTable>` " +
+        "SELECT DISTINCT <rowId> FROM `<project>.<dataset>.<softDeleteExtTable>`";
+
+    /**
+     * Insert row ids into the corresponding soft delete table for each table provided.
+     *
+     * @param dataset repo dataset that we are deleting data from
+     * @param tables list of tables that should have corresponding external tables with row ids to soft delete
+     * @param suffix a bq-safe version of the flight id to prevent different flights from stepping on each other
+     */
+    public TableResult applySoftDeletes(Dataset dataset,
+                                 List<DataDeletionTableModel> tables,
+                                 String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+
+        // we want this soft delete operation to be atomic, we we will combine all of the inserts into one statement
+        // that we send to bigquery.
```

Comment:
Could you elaborate more on the "atomic" part of the comment please? I don't think running a SQL script or multiple statements separated by semicolons makes any additional guarantees when compared with running the statements individually, though I'm not familiar with BigQuery's behavior specifically. Do you mean that you just want to make a single BQ network request?

Reply:
This combines them all into one BigQuery `Job`. I was hoping that this would treat them as a transaction that succeeds or fails atomically. On further investigation it looks like [BigQuery creates a child job for each statement](https://cloud.google.com/bigquery/docs/reference/standard-sql/scripting#bigquery-scripting), so my hopes are dashed! I'll update the comment and make a ticket to revisit this. I think we need something along the lines of a soft delete tag to be able to get transaction like behavior in the case of a failure in one of these child jobs.

Reply:
Making a separate ticket to address this sounds good. Especially since soft-deleting the same row twice is not allowed, so users can't just keep executing the same request until it all succeeds.

Reply:
https://broadworkbench.atlassian.net/browse/DR-910